### PR TITLE
Fixes #33425 - Optionally prefer registered through proxy for REX

### PR DESCRIPTION
**Pull request by adamruzicka**

### CODE REVIEW

**Comment by jlsherrill** [comment]
[test katello]

**Comment by jturel** [comment]
ping @adamruzicka - rebase needed :)

**Comment by jturel** [comment]
@adamruzicka I need some help understanding / evaluating this change. Maybe the content_source would be a better choice than registered through? Was the "Registered Through" smart proxy chosen because that's something that the user or support saw in the UI?

Regarding the new setting - it seems necessary but if possible I think we should avoid it (I'm thinking of reducing complexity, code paths, making debugging easier). Can you think of a way we can do without it? I guess making the strategy you added be the default would not be good?

**Comment by adamruzicka** [comment]
> Was the "Registered Through" smart proxy chosen because that's something that the user or support saw in the UI?

It felt a lot like people didn't understand how rex picks a proxy and were hitting our support with "the host is registered to that proxy, why isn't it using that proxy for rex?"

> Can you think of a way we can do without it? I guess making the strategy you added be the default would not be good?

Not really, it would be a bad default. From my point of view this is implemented purely for people who can't be bothered to set things up properly.

File: app/models/katello/concerns/host_managed_extensions.rb [hunk 1]
```diff
@@ -28,6 +28,16 @@ def update_os_from_facts
             content_facet.kickstart_repository_id = valid_repos.first
           end
         end
+
+        def remote_execution_proxies(provider, *_rest)
```
Comment (jturel) [inline app/models/katello/concerns/host_managed_extensions.rb:32]: This method doesn't seem to be getting called (my breakpoint isn't hit when kicking off a REX job)

File: lib/katello/engine.rb [hunk 2]
```diff
@@ -226,6 +226,10 @@ class Engine < ::Rails::Engine
       ::AuditSearch::ClassMethods.prepend Katello::Concerns::AuditSearch
       ::Widget.singleton_class.prepend Katello::Concerns::WidgetExtensions::ClassMethods
 
+      if with_remote_execution?
```
Comment (jturel) [inline lib/katello/engine.rb:229]: This is throwing an error - it should be Katello.with_remote_execution

**Comment by jturel** [comment]
I understand the reasoning now but it doesn't seem to work. Even with the setting set to true the default smart proxy is getting selected rather than the registered_through proxy which I've configured for my content host

File: lib/katello/engine.rb (hunk 2 above)
Comment (adamruzicka) [reply lib/katello/engine.rb]: Right, not sure how I managed to let this slip

File: app/models/katello/concerns/host_managed_extensions.rb (hunk 1 above)
Comment (adamruzicka) [reply app/models/katello/concerns/host_managed_extensions.rb:32]: Not sure what your environment looks like, but if it is a hijacked production setup, then all the dynflow-sidekiq@* services need to be cycled to pick up the changes.

Comment (jturel) [reply app/models/katello/concerns/host_managed_extensions.rb:32]: Just the usual katello devel environment - centos7-katello-devel.

I guess it's that these extensions are prepended first, and then the REX extensions are prepended and effectively overriding this method?

::Host.first.remote_execution_proxies(nil) isn't hitting this method in the console :(

**Comment by jturel** [comment]
[test katello]

**Review by jturel — APPROVED** [review]
ACK!

I was able to toggle the setting and see the action move between my default and external smart proxies. Thanks @adamruzicka

**Comment by jturel** [comment]
@adamruzicka rubocop :(

**Pull request merged**